homeassistant/components/skybell/__init__.py

```diff
@@ -1,88 +1,172 @@
     CONF_PASSWORD,
     CONF_USERNAME,
-    __version__,
 )
-import homeassistant.helpers.config_validation as cv
-from homeassistant.helpers.entity import Entity
-
-_LOGGER = logging.getLogger(__name__)
-
-ATTRIBUTION = "Data provided by Skybell.com"
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv, device_registry as dr
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
 
-NOTIFICATION_ID = "skybell_notification"
-NOTIFICATION_TITLE = "Skybell Sensor Setup"
+from .const import (
+    AGENT_IDENTIFIER,
+    ATTRIBUTION,
+    DATA_COORDINATOR,
+    DATA_DEVICES,
+    DEFAULT_CACHEDB,
+    DEFAULT_NAME,
+    DOMAIN,
+    MIN_TIME_BETWEEN_UPDATES,
+)
 
-DOMAIN = "skybell"
-DEFAULT_CACHEDB = "./skybell_cache.pickle"
-DEFAULT_ENTITY_NAMESPACE = "skybell"
+_LOGGER = getLogger(__name__)
 
-AGENT_IDENTIFIER = f"HomeAssistant/{__version__}"
+PLATFORMS = [BINARY_SENSOR, CAMERA, LIGHT, SENSOR, SWITCH]
 
 CONFIG_SCHEMA = vol.Schema(
-    {
-        DOMAIN: vol.Schema(
-            {
-                vol.Required(CONF_USERNAME): cv.string,
-                vol.Required(CONF_PASSWORD): cv.string,
-            }
-        )
-    },
+    vol.All(
+        # Deprecated in Home Assistant 2021.9
+        cv.deprecated(DOMAIN),
+        {
+            DOMAIN: vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME): cv.string,
+                    vol.Required(CONF_PASSWORD): cv.string,
+                }
+            )
+        },
+    ),
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass, config):
-    """Set up the Skybell component."""
-    conf = config[DOMAIN]
-    username = conf.get(CONF_USERNAME)
-    password = conf.get(CONF_PASSWORD)
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the SkyBell component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    entry_config = {}
+    if DOMAIN in config:
+        for parameter in config[DOMAIN]:
+            if parameter == CONF_USERNAME:
+                entry_config[CONF_EMAIL] = config[DOMAIN][parameter]
+            else:
+                entry_config[parameter] = config[DOMAIN][parameter]
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN,
+                    context={"source": SOURCE_IMPORT},
+                    data=entry_config,
+                )
+            )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Skybell from a config entry."""
+    email = entry.data[CONF_EMAIL]
+    password = entry.data[CONF_PASSWORD]
 
+    api = Skybell(
+        cache_path=hass.config.path(DEFAULT_CACHEDB),
+        agent_identifier=AGENT_IDENTIFIER,
+    )
```

Comment:
Can this throw?

Reply:
I guess it's possible when we try `dump` in `pickle` during saving cache but it seems very unlikely. We are not getting devices, nor are we logging in here. We are simply loading the cache if it exists and errors are not thrown if not found. We just move to login inside the try after that.